Extract controller features that update the operator's status into a new internal package

qbarrand opened the pull request:
Add unit tests and mocks for those types.

/cc @yevgeny-shnaidman

qbarrand commented:
/ok-to-test

pmtk commented on internal/controllers/state/clusteroperator.go:
```diff
@@ -0,0 +1,139 @@
+// PUBLIC
+// SpecialResourcesStatusInit Depending on what error we're getting from the
+// reconciliation loop we're updating the status
+// nil -> All things good and default conditions can be applied
```
This does not seem to adhere to Go docs guidelines. And do we need "PUBLIC" and "PRIVATE" comments?

qbarrand replied:
Sorry, this is a leftover of me trying to figure out which methods needed to be included in the public interface. I will review all the function comments.

pmtk commented on internal/controllers/state/clusteroperator.go:
```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+// GetOrCreate refreshes the locally cached configv1.ClusterOperator instance with what is currently in Kubernetes.
```
Nit: `in Kubernetes` -> `in cluster`? WDYT

qbarrand replied:
Absolutely.

pmtk reviewed (commented):
Just these two low prio comments. Code looks good to me

qbarrand committed:
Add ClusterOperatorManager and the StatusUpdater.

Extract those features from the controller package.
Unit test public methods of the types.
Add mocks.

qbarrand commented:
/ok-to-test

ybettan commented:
/hold

ybettan commented:
/unhold

ybettan commented:
/hold
Tide is missing

qbarrand commented:
/unhold tide was missing because of GitHub rate limiting.

qbarrand commented:
/unhold

qbarrand commented:
/ok-to-test

yevgeny-shnaidman commented on internal/controllers/state/clusteroperator.go:
```diff
@@ -81,26 +82,26 @@ func (com *clusterOperatorManager) Refresh(ctx context.Context, cond []configv1.
-	if err != nil {
+	co := &configv1.ClusterOperator{}
+
+	if err := com.kubeClient.Get(ctx, types.NamespacedName{Name: com.operatorName}, co); err != nil {
```
what do you think about leaving the API as is, and just changing the inner implementation? I think that the code will be a little bit clearer, no?

qbarrand replied:
We need to group those functions into new types to be able to mock the Kubernetes client. 
The name changes do make things clearer in my opinion, e.g. I think that `clusterOperatorManager.Refresh()` is more obvious than `SpecialResourcesStatus()`.
I understand that it can make the review more painful though. I can revert to the old names, but I would really change them in the long run, as they can be quite confusing. Or did I miss your point?

yevgeny-shnaidman replied:
The new names for clusterOepratorManager are great, i was talking about keeping the old API names for the clients package, and just changing the inner implementation to runtime-client. For example,  kubeClient.ClusterOperatorGet sounds more clear to me then kubeClient.Get (with namespaces name). WDYT?

qbarrand replied:
I removed that commit from the PR. Let us things related to the Kubernetes client as they are for now while we investigate if we can come up with a better `pkg/clients` package.
@yevgeny-shnaidman PTAL at the rest of the PR 🙂

yevgeny-shnaidman commented on controllers/specialresource.go:
```diff
@@ -21,11 +21,6 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-// GetName of the special resource operator
```
i am probably missing something, but where is the new implementation of this function

qbarrand replied:
There is no new implementation of that function. It was only used to name the `ClusterOperator` resource; it is now the `operatorName` field of the `clusterOperatorManager` struct.

qbarrand commented:
/ok-to-test

yevgeny-shnaidman commented:
/lgtm
